utils/stable_matching.ml

```diff
@@ -0,0 +1,372 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                Malo Monin, projet Cambium, Inria Paris                 *)
+(*                                                                        *)
+(*   Copyright 2024 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type layer = { left_candidates: int list; pref:int }
+
+type ('a,'v) matches = {
+  left : 'a list;
+  pairs : ('v * 'v) list;
+  right : 'a list;
+}
+
+module Item = struct
+  type ('v, 'k) t = {
+    name: string;
+    item: 'v;
+    kind: 'k;
+  }
+
+  let name f = f.name
+  let item f = f.item
+  let kind i = i.kind
+end
+type nonrec ('v,'k) item_matches =  (('v,'k) Item.t, 'v) matches
+
+
+(** An implementation (in [diff]) of Zoltan Kiraly's "New Algorithm," presented
+    in "Linear Time Local Approximation Algorithm for Maximum Stable Marriage":
+    https://www.mdpi.com/1999-4893/6/3/471. It computes a 3/2-approximation of
+    a maximum stable marriage in linear time (linear in the sum of the lengths
+    of the preference lists). *)
```

Comment:
Are there reasonable test vectors one could use to get confidence in the stable-matching algorithm? Are they exercised in the testsuite, or would it be worth having dedicated unit tests just for the stable matching logic?

Reply:
I have added an implementation of Gale-Shapley and the corresponding property tests checking that
-  the matching produced are at least weakly stable
-  Gale-Shapley and Kirarly's algorithm produce the same matching for instance with complete preferences without ties.
- Kiraly's algorithm produces larger and stabler matching in average.